### PR TITLE
[#9057] FeedbackConstSumOptionQuestionUiTest failing on live server

### LINKED ISSUE #9057: FeedbackConstSumOptionQuestionUiTest failing on live server
v6.10.0

I'm running tests in Windows, against the live server

java.lang.AssertionError: 
Expected :["Choice 3","Choice 1","Choice 2","Choice 4"]
Actual   :["Choice 1","Choice 2","Choice 3","Choice 4"]
 <Click to see difference>

	at org.testng.AssertJUnit.assertEquals(AssertJUnit.java:101)
	at org.testng.AssertJUnit.assertEquals(AssertJUnit.java:108)
	at teammates.test.cases.BaseTestCase.assertEquals(BaseTestCase.java:143)
	at teammates.test.cases.browsertests.FeedbackConstSumOptionQuestionUiTest.testReorderOptions(FeedbackConstSumOptionQuestionUiTest.java:285)
	at teammates.test.cases.browsertests.FeedbackConstSumOptionQuestionUiTest.testEditPage(FeedbackConstSumOptionQuestionUiTest.java:58)
	at teammates.test.cases.browsertests.FeedbackConstSumOptionQuestionUiTest.allTests(FeedbackConstSumOptionQuestionUiTest.java:45)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:659)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:845)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1153)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:771)
	at org.testng.TestRunner.run(TestRunner.java:621)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1199)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1124)
	at org.testng.TestNG.run(TestNG.java:1032)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)

### PR DESCRIPTION
**Pull request by mg14777**
Fixes #9057 

**Outline of Solution**

@xpdavid @damithc Can someone try this fix to see if it works. So far I have been able to reproduce the error on certain occasions on resizing my window. All the tests were run on Firefox 43.0.1. I tried two different approaches.

In the first one I tried to use explicit wait where i move the element to the target location and wait until a placeholder element is created after which i release it. This didn't seem to work every time though.

Explicit wait:
drag_drop_wait

The second approach was similar to the original technique i was using by breaking down the movement into smaller movements. This time I have written it in a more concise way though. This technique seems to work with both reduced and original window size.

Move breakdown:
drag_drop_2_movements

As for the menubar overlaying the target position, I tried to recreate the situation but the window automatically scrolls to ensure the target element is in focus as per the `moveToElement` function provided by selenium.

### CODE REVIEW

**Comment by LiHaoTan** [comment]
@mg14777 is there a reason why `Actions#dragAndDrop` is not used?

**Comment by mg14777** [comment]
@LiHaoTan I tried using that while working on the original PR but it doesn't work. Details [here](https://github.com/TEAMMATES/teammates/pull/8889#issuecomment-397766993)

**Comment by xpdavid** [comment]
I remember somewhere in the code base stated that Action is not recommended because of some defects in drivers. I think here `scrollElementToCenterAndClick`. Can take a look to see if we can upgrade the driver to solve the problem.

**Comment by mg14777** [comment]
Based on my experience there wasn't a problem with the actual `dragAndDrop` method itself (although a lot of people have pointed out [issues](https://github.com/SeleniumHQ/selenium/issues?q=is%3Aissue+drag+is%3Aopen) with it) provided by Selenium. It was more about the execution speed of the drag movement which is too fast and doesn't give enough time for the grid to adjust as shown below. I tried to look if there's a way to slow it down but it seems to be a requested feature atm
[https://github.com/SeleniumHQ/selenium/issues/2291](https://github.com/SeleniumHQ/selenium/issues/2291)
That's why i finally resorted to using an Actions chain approach as given here
[https://github.com/SeleniumHQ/selenium/wiki/Advanced-User-Interactions#generating-action-chains](https://github.com/SeleniumHQ/selenium/wiki/Advanced-User-Interactions#generating-action-chains)

drag_drop_method_fast_1

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java [hunk 1]
```diff
@@ -1475,8 +1475,10 @@ public void dragAndDropQuestionOption(String qnType, int qnNumber, int sourceInd
                 + targetIndex + "-" + qnNumber + "']//span[@class='glyphicon glyphicon-resize-vertical']"));
 
         Actions builder = new Actions(browser.driver);
-        // drag option 10 units above target and release
+        // drag option 10 units below target, move to target, move 10 units above target and release
```
Comment (xpdavid) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java]: Can you provide some comments regarding why we do the extrat stuffs?

**Comment by xpdavid** [comment]
@mg14777 Any updates? Can you check if the current Appveyor fail for the same reason? https://ci.appveyor.com/project/damithc/teammates/branch/master

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java (hunk 1 above)
Comment (mg14777) [reply src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java]: @xpdavid Really sorry for such a late response. Been quite busy lately. I have added comments explaining the extra movements.

Actually I tried testing different combinations again this morning and realized that using 2 movements still works. Selenium 3 has added `moveInTicks` functionality [https://github.com/SeleniumHQ/selenium/blob/dbe5d01c93b54958c3955b36f5166a3f704b74c1/java/client/src/org/openqa/selenium/interactions/Actions.java#L413](https://github.com/SeleniumHQ/selenium/blob/dbe5d01c93b54958c3955b36f5166a3f704b74c1/java/client/src/org/openqa/selenium/interactions/Actions.java#L413) that adds a duration of 100 milisecond. It should hopefully solve this issue.

**Comment by mg14777** [comment]
> Can you check if the current Appveyor fail for the same reason? https://ci.appveyor.com/project/damithc/teammates/branch/master

@xpdavid Yes some tests in Appveyor are failing because of drag and drop. Please see my comment above.

**Pull request merged**